### ISSUE TITLE
Test My Data instead of MNIST files

### ISSUE THREAD

**Issue by RoboEvangelist** [ageron/handson-ml#183]
Hi,
Excellent book. One of the best out there.

My questions is, how can I replace the MNIST images in the "High Accuracy CNN for MNIST" with my own dataset? I have 18000 images (9000 images per class, so two classes), and I want to feed that into the same network. My images are HD, grayscale.
How can I feed my own images? Any example you have with the same network architecture?

Thanks,

**Comment by ageron** [comment]
Hi @RoboEvangelist ,

Unfortunately, 9000 images per class is not very much, especially with HD images.  You probably want to resize the images down to the smallest size that you can, as long a human can still perform the classification task.  You can also try cropping the images as much as possible.  Basically, reduce the dimensionality as much as you can.

Then you can use mostly the same code to train your model, just change the height and width appropriately:

```python
height = 28
width = 28
```

The next thing you need to change is this line:

```python
pool3_flat = tf.reshape(pool3, shape=[-1, pool3_fmaps * 14 * 14])
```

The 14*14 should be replaced with some other value that depends on the image height and width, probably just height//2 * width//2, but it might be slightly off by one or two pixels because of rounding errors. The easiest way to know is to print `pool3.shape`.

Next, if you had any number of classes equal or greater than 3, then you would just change the line `n_outputs` to that number of classes, and that would be all. In your case, since it's binary classification, you can just have a single output neuron for the positive class (you could just set `n_outputs` to be equal to 2, but it's more efficient to set it to 1 instead).  So set `n_outputs = 1` and replace `tf.nn.softmax` with `tf.nn.sigmoid`:

```python
Y_proba = tf.nn.sigmoid(logits, name="Y_proba")
```

Then the cross entropy can be computed like this:

```python
xentropy = tf.nn.sigmoid_cross_entropy_with_logits(logits=logits, labels=y)
```

The labels must be float32 instead of int32: 0.0 for negative instances, and 1.0 for positive instances.
Then you must change the execution phase to create batches of the right shape and type, and you're done.

However... What kinds of images are they? If they look like the images in the ImageNet dataset (i.e., random pictures you could take with your phone), then you will probably get the best performance by using a pretrained convolutional neural network such as Inception-v3, and fine-tune it with your dataset. More generally, if you can find a neural net that was trained on a larger dataset of images that look more or less like yours, it would be best.  Check out the solution to exercise 9 in [chapter 13](https://github.com/ageron/handson-ml/blob/master/13_convolutional_neural_networks.ipynb).

Hope this helps,
Aurélien

**Comment by RoboEvangelist** [comment]
my images are HD grayscale and they come from an near-infrared camera, and the background does not change much. The object that I am trying to detect tends to cover most of the image on every picture that I have (so that helps a lot).

I could use inception V3, but I noticed the tensorflow version is a little big slow (i want something like 30 FPS of detection rate if possible).
I will try to add the changes you suggested to the chapter 9 code.

I'll try to record something like 30000 per class, since it is possible for me to do it. But I thought I'd start with a small dataset.

Thanks for the guide.
